### PR TITLE
Fix CBMC proof for DHCP

### PR DESCRIPTION
**Pull request by tony-josi-aws**
This PR fixes the build issues and proof failures for the DHCP CBMC proofs post the IPv6 changes.

Test Steps
-----------
```
cd test/cbmc/proofs/
python run-cbmc-proofs.py
```
Related Issue
-----------
Failing CBMC proofs

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c [hunk 1]
```diff
@@ -42,40 +42,114 @@
...
+    pxNetworkBuffer->pucEthernetBuffer = ( (uint8_t *) malloc( xRequestedSizeBytes + ( ipIP_TYPE_OFFSET) )) + ipIP_TYPE_OFFSET;
+    __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer != NULL );
+
+    pxNetworkBuffer->xDataLength = xRequestedSizeBytes;
+    return pxNetworkBuffer;
+}
+
+void FreeRTOS_ReleaseUDPPayloadBuffer( void * pvBuffer )
+{
+    __CPROVER_assert( pvBuffer != NULL,
```
[the remark this answers is not in the document]
Comment (tony-josi-aws) [reply test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c:121]: For this proof `pvBuffer` can be allocated either by `pxGetNetworkBufferWithDescriptor` or stubbed `FreeRTOS_recvfrom` (zero copy version), both of them does allocation differently with different ptr offsets.

File: test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c [hunk 2]
```diff
@@ -42,40 +42,114 @@
...
 /****************************************************************
 * The proof of vDHCPProcess
 ****************************************************************/
 
 void harness()
 {
     BaseType_t xReset;
-    eDHCPState_t eExpectedState;
+    BaseType_t xDoCheck;
+    //eDHCPState_t eExpectedState;
```
[the remark this answers is not in the document]
Comment (tony-josi-aws) [reply test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c]: Fixed

File: source/FreeRTOS_DHCP.c [hunk 3]
```diff
@@ -287,7 +287,7 @@
                     /* PAss the address of a pointer pucUDPPayload, because zero-copy is used. */
                     lBytes = FreeRTOS_recvfrom( xDHCPv4Socket, &( pucUDPPayload ), 0, FREERTOS_ZERO_COPY, NULL, NULL );
 
-                    if( lBytes > 0 )
+                    if( lBytes > 0 && pucUDPPayload != NULL )
```
[the remark this answers is not in the document]
Comment (tony-josi-aws) [reply source/FreeRTOS_DHCP.c]: Fixed all format issues

File: test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c [hunk 4]
```diff
@@ -42,40 +42,114 @@
...
+{
+    return pdTRUE;
+}
+
+BaseType_t vSocketBind( FreeRTOS_Socket_t * pxSocket,
+                        struct freertos_sockaddr * pxBindAddress,
+                        size_t uxAddressLength,
+                        BaseType_t xInternal )
+{
+    return 0;
```
[the remark this answers is not in the document]
Comment (tony-josi-aws) [reply test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c]: Fixed

File: test/cbmc/patches/FreeRTOSIPConfig.h [hunk 5]
```diff
@@ -135,7 +135,9 @@ extern uint32_t ulRand();
  * reason.  The static configuration used is that passed into the stack by the
  * FreeRTOS_IPInit() function call. */
 #define ipconfigUSE_DHCP                         1
-#define ipconfigDHCP_REGISTER_HOSTNAME           1
+#ifndef ipconfigDHCP_REGISTER_HOSTNAME
+    #define ipconfigDHCP_REGISTER_HOSTNAME           1
+#endif
```
Comment (AniruddhaKanhere) [inline test/cbmc/patches/FreeRTOSIPConfig.h:140]: Why do we need this?

File: test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c [hunk 6]
```diff
@@ -42,40 +42,120 @@
...
-BaseType_t prvProcessDHCPReplies( BaseType_t xExpectedMessageType )
+BaseType_t __CPROVER_file_local_FreeRTOS_DHCP_c_prvProcessDHCPReplies( BaseType_t xExpectedMessageType,
+                                             NetworkEndPoint_t * pxEndPoint )
 {
     return nondet_BaseType();
 }
 
+BaseType_t xSocketValid( const ConstSocket_t xSocket )
+{
+    return pdTRUE;
```
Comment (AniruddhaKanhere) [inline test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c:82]: Should this always return a `pdTRUE`?

Comment (tony-josi-aws) [reply test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c:84]: This is done to get rid of asserts in the DHCP source

File: test/cbmc/patches/FreeRTOSIPConfig.h (hunk 5 above)
Comment (tony-josi-aws) [reply test/cbmc/patches/FreeRTOSIPConfig.h]: resolved

**Pull request merged**